File: examples/java/java-booking-03_preserve_whole_object/src/main/java/tv/codely/booking/Booking.java
package tv.codely.booking;

import java.time.LocalDateTime;

public final class Booking {
    private final BookingId     id;
    private final LocalDateTime startDate;
    private final LocalDateTime endDate;
    private final Customer      customer;
    private final BookingType   bookingType;
    private final Discount      discount;
    private final Tax           tax;

    public Booking(
        BookingId id,
        DateRange bookingDateRange,
        Customer customer,
        BookingType bookingType,
        Discount discount,
        Tax tax
    ) {
        this.id          = id;
        this.startDate   = bookingDateRange.startDate();
        this.endDate     = bookingDateRange.endDate();
        this.customer    = customer;
        this.bookingType = bookingType;
        this.discount    = discount;
        this.tax         = tax;
    }

    public BookingStatus statusFor(LocalDateTime date) {
        if (date.isBefore(startDate)) {
            return BookingStatus.NOT_STARTED;
        }

        if (date.isAfter(startDate) && date.isBefore(endDate)) {
            return BookingStatus.ACTIVE;
        }

        return BookingStatus.FINISHED;
    }
}

File: examples/csharp/csharp-booking-04_tell_dont_ask/src/Booking/Booking.cs
using System;

namespace CodelyTv.Booking
{
    public sealed class Booking
    {
        private readonly BookingId id;
        private readonly DateRange dateRange;
        private readonly Customer customer;
        private readonly BookingType bookingType;
        private readonly Discount discount;
        private readonly Tax tax;

        public Booking(
            BookingId id,
            DateRange dateRange,
            Customer customer,
            BookingType bookingType,
            Discount discount,
            Tax tax
        )
        {
            this.id = id;
            this.dateRange = dateRange;
            this.customer = customer;
            this.bookingType = bookingType;
            this.discount = discount;
            this.tax = tax;
        }

        public BookingStatus StatusFor(DateTime date)
        {
            if (dateRange.IsAfter(date))
            {
                return BookingStatus.NOT_STARTED;
            }

            if (dateRange.IsBetween(date))
            {
                return BookingStatus.ACTIVE;
            }

            return BookingStatus.FINISHED;
        }
    }
}
